### PR TITLE
Add support for parameter UI widgets

### LINKED ISSUE #88: Support for parameter UI widgets (YAML)
Support in YAML to ask for a custom UI widget for parameter editing like SQL or DatumAlias.

Example UI:
Image

Example YAML:
```
 - step:
     name: snowflake-query
     image: valohai/dbconnectors
     command: python query.py {parameters}
     icon: https://valohai.com/img/orig/SNOW-35164165.png
     category: Database query
     parameters:
     - name: SQLquery
       type: string
       default: SELECT * FROM table
       widget: sql
     - name: datum_alias
       type: string
       widget: datumalias
```

### PR DESCRIPTION
**Pull request by JuhaKiili**
Example YAML:
```
 - step:
     name: snowflake-query
     image: valohai/dbconnectors
     command: python query.py {parameters}
     icon: https://valohai.com/img/orig/SNOW-35164165.png
     category: Database query
     parameters:
     - name: SQLquery
       type: string
       default: SELECT * FROM table
       widget: sql
     - name: datum_alias
       type: string
       widget: 
         type: datumalias
         settings:
           width: 100
```

Fixes #88

### CODE REVIEW

File: valohai_yaml/schema/param-item.yaml [hunk 1]
```diff
@@ -78,3 +78,12 @@ properties:
     description: |
       The separator to use when interpolating multiple values into a command parameter.
       Defaults to a comma.
+  widget:
+    type: string
+    default: none
+    enum:
+      - none
+      - sql
+      - datumalias
```
Comment (akx) [inline valohai_yaml/schema/param-item.yaml:87]: Let's not lock the given values down here? It's always a pain in the rear to have to update the whole ecosystem depending on this library.
```suggestion
```

Also, I think we should look into the future a bit and allow this to be an object as well as a string:

```yaml
widget: sql
```
would be equivalent to
```yaml
widget:
  type: sql
```
and in the future each widget might have other settings too?

Parsing would always result in a dict or None, serializing would attempt the shortest form, i.e. if the dict in the Step object has `len(widget) == 1 and widget.get("type")`, then it'd be serialized back to a string.

**Review by akx — CHANGES_REQUESTED** [review]

File: valohai_yaml/objs/parameter_widget.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    def cast(cls, value: Optional[Union['ParameterWidget', OrderedDict, dict, str]]) -> Optional['ParameterWidget']:
+        if not value:
+            return None
+        if isinstance(value, (dict, OrderedDict)):
+            return cls(type=value['type'])
+        if isinstance(value, cls):
+            return value
+        value = str(value).lower()
+        if value == 'none':
+            return None
```
Comment (akx) [inline valohai_yaml/objs/parameter_widget.py:27]: I wonder if the special `"none"` value is really necessary at all..?

File: tests/test_step_parsing.py [hunk 3]
```diff
@@ -154,3 +156,10 @@ def test_timeouts(timeouts_config):
 def test_bling(example1_config: Config) -> None:
     assert example1_config.steps['run training'].category == "Training"
     assert example1_config.steps['run training'].icon == "https://valohai.com/assets/img/valohai-logo.svg"
+
+
+def test_widget(example1_config: Config) -> None:
+    assert example1_config.steps['run training'].parameters['sql-query'].widget and \
+        example1_config.steps['run training'].parameters['sql-query'].widget.type == "sql"
+    assert example1_config.steps['run training'].parameters['output-alias'].widget and \
+        example1_config.steps['run training'].parameters['output-alias'].widget.type == "datumalias"
```
Comment (akx) [inline tests/test_step_parsing.py:165]: DRY... :)

```suggestion
    parameters = example1_config.steps['run training'].parameters
    assert parameters['sql-query'].widget and parameters['sql-query'].widget.type == "sql"
    assert parameters['output-alias'].widget and parameters['output-alias'].widget.type == "datumalias"
```

Comment (akx) [inline tests/test_step_parsing.py:165]: Can you also check that `widget` is `None` as expected for some other parameter?

File: valohai_yaml/objs/parameter_widget.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    def __init__(
+        self,
+        *,
+        type: str,
+    ) -> None:
+        self.type = type.lower()
+
+    @classmethod
+    def cast(cls, value: Optional[Union['ParameterWidget', OrderedDict, dict, str]]) -> Optional['ParameterWidget']:
```
Comment (akx) [inline valohai_yaml/objs/parameter_widget.py:18]: Also, ah, `cast` is only a thing for our `Enum`s; this should be `parse()` (and you should also define `serialize()`, so ParameterWidgets with just a type can be serialized back to a simple `str`.

**Review by akx — CHANGES_REQUESTED** [review]

File: valohai_yaml/objs/parameter_widget.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    """An UI widget for editing parameter values."""
+
+    def __init__(
+        self,
+        *,
+        type: str,
+        settings: Optional[dict] = None,
+    ) -> None:
+        self.type = type.lower()
+        self.settings = settings
```
Comment (akx) [inline valohai_yaml/objs/parameter_widget.py:17]: Might be nicer that `self.settings` always was a thing, even if empty.

```suggestion
        self.settings = dict(settings or {})
```

This doesn't need changes to the serializer though since empty dicts are falsy ✨

File: valohai_yaml/objs/parameter_widget.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+                'type': self.type,
+                'settings': self.settings,
+            }
+        return self.type
+
+    @classmethod
+    def parse(cls, data: Union['ParameterWidget', dict, str]) -> 'ParameterWidget':
+        if isinstance(data, dict):
+            return cls(
+                type=data.get('type', ''),
```
Comment (akx) [inline valohai_yaml/objs/parameter_widget.py:31]: I think `type` is required?
```suggestion
                type=data['type'],
```

File: valohai_yaml/objs/parameter_widget.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+                settings=data.get('settings'),
+            )
+
+        if isinstance(data, cls):
+            return data
+
+        value = str(data).lower()
```
Comment (akx) [inline valohai_yaml/objs/parameter_widget.py:38]: What's the reason for the lower-casing? And why isn't it done for the other style of `type`?

File: valohai_yaml/schema/param-item.yaml [hunk 8]
```diff
@@ -1,3 +1,4 @@
+# Here is the full schema
```
Comment (akx) [inline valohai_yaml/schema/param-item.yaml:1]: I'm not sure this comment has much value..?

File: valohai_yaml/schema/param-item.yaml [hunk 9]
```diff
@@ -78,3 +79,17 @@ properties:
...
       Defaults to a comma.
+  widget:
+    oneOf:
+      - type: object
+        required:
+        - type
+        properties:
+          type:
+            type: string
+            default: none
```
Comment (akx) [inline valohai_yaml/schema/param-item.yaml:90]: I don't think there should be a default..?
```suggestion
```

**Review by akx — CHANGES_REQUESTED** [review]

File: valohai_yaml/schema/param-item.yaml (hunk 8 above)
Comment (JuhaKiili) [reply valohai_yaml/schema/param-item.yaml]: Leftover from my dialog with ChatGPT ;)

File: valohai_yaml/objs/parameter_widget.py (hunk 7 above)
Comment (JuhaKiili) [reply valohai_yaml/objs/parameter_widget.py:38]: Not doing it for the dict was my error. Pushed a fix now.

The reasoning for lowercase was that people will use `SQL`, `Sql` and `sql` and I don't see the upside of making this case sensitive.

Comment (akx) [reply valohai_yaml/objs/parameter_widget.py:38]: Fair enough.

File: valohai_yaml/objs/parameter_widget.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+            return cls(
+                type=str(data['type']).lower(),
+                settings=data.get('settings'),
+            )
+
+        if isinstance(data, cls):
+            return data
+
+        value = str(data).lower()
+        return cls(type=value)
```
Comment (akx) [inline valohai_yaml/objs/parameter_widget.py:39]: Since you're `.lower`ing this in the constructor anyway,
```suggestion
        return cls(type=data)
```

File: valohai_yaml/objs/parameter_widget.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+                'type': self.type,
+                'settings': self.settings,
+            }
+        return self.type
+
+    @classmethod
+    def parse(cls, data: Union['ParameterWidget', dict, str]) -> 'ParameterWidget':
+        if isinstance(data, dict):
+            return cls(
+                type=str(data['type']).lower(),
```
Comment (akx) [inline valohai_yaml/objs/parameter_widget.py:31]: You're `.lower`ing the type in the constructor anyway, so
```suggestion
                type=data['type'],
```

File: valohai_yaml/objs/parameter_widget.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+class ParameterWidget(Item):
+    """An UI widget for editing parameter values."""
+
+    def __init__(
+        self,
+        *,
+        type: str,
+        settings: Optional[dict] = None,
+    ) -> None:
+        self.type = type.lower()
```
Comment (akx) [inline valohai_yaml/objs/parameter_widget.py:16]:
```suggestion
        self.type = str(type).lower()
```

File: valohai_yaml/objs/parameter.py [hunk 13]
```diff
@@ -199,11 +203,16 @@ def lint(
         lint_result: LintResult,
         context: LintContext,
     ) -> None:
+        prefix = f'Step {context["step"].name}, parameter {self.name}'
+        self.lint_flags(prefix, lint_result)
+        self.lint_default(prefix, lint_result)
```
Comment (akx) [inline valohai_yaml/objs/parameter.py:208]: What's the reason for this change/refactoring, by the way? I don't see it described in the commit message or PR description...

**Review by akx — CHANGES_REQUESTED** [review]

File: valohai_yaml/objs/parameter.py [hunk 14]
```diff
@@ -62,6 +64,8 @@ def __init__(
         self.choices = (list(choices) if choices else None)
         self.multiple = MultipleMode.cast(multiple)
         self.multiple_separator = str(multiple_separator or ',')
+        self.widget = ParameterWidget.parse(widget)
```
Comment (akx) [inline valohai_yaml/objs/parameter.py:67]: Just
```suggestion
        self.widget = widget
```
here, and call `ParameterWidget.parse` in `Parameter`'s `parse`, á la

```python
    @classmethod
    def parse(cls, data: SerializedDict) -> 'Parameter':
        kwargs = data.copy()
        kwargs['widget'] = ParameterWidget.parse(data.get('widget'))
        return cls(**kwargs)
```

**Review by akx — CHANGES_REQUESTED** [review]

File: valohai_yaml/objs/parameter.py (hunk 14 above)
Comment (JuhaKiili) [reply valohai_yaml/objs/parameter.py]: Sorry! I was rushing it and had a lukihäiriö. 

It should be kosher now.

File: valohai_yaml/objs/parameter_widget.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+    def serialize(self) -> Optional[Union[SerializedDict, str]]:
+        if self.settings:
+            return {
+                'type': self.type,
+                'settings': self.settings,
+            }
+        return self.type
+
+    @classmethod
+    def parse(cls, data: Optional[Union['ParameterWidget', dict, str]]) -> 'ParameterWidget':
```
Comment (akx) [inline valohai_yaml/objs/parameter_widget.py:28]: Based on the code (I swear it had `if data is None` in an earlier revision?!) `Optional` is no longer true here; this would do the wrong thing for `None`.

Either get rid of that –

```suggestion
    def parse(cls, data: Union['ParameterWidget', dict, str]) -> 'ParameterWidget':
```

or alternately make it return `None` for falsy values...

```suggestion
    def parse(cls, data: Optional[Union['ParameterWidget', dict, str]]) -> Optional['ParameterWidget']:
        if not data:
            return None
```

I'd go with option 2, because it would simplify `Parameter.parse` too; you wouldn't need to check whether `widget` is in `data`...

**Review by akx — CHANGES_REQUESTED** [review]
Sorry – one more tweak...

File: valohai_yaml/objs/parameter_widget.py (hunk 15 above)
Comment (JuhaKiili) [reply valohai_yaml/objs/parameter_widget.py]: I was going for option 2 earlier, but the mypy error made me think its not cool (?)

```
Return type "Optional[ParameterWidget]" of "parse" incompatible with return type "ParameterWidget" in supertype "Item"  [override]
```

I've pushed option 1 as separate commit now.

Comment (akx) [reply valohai_yaml/objs/parameter_widget.py]: Alright! I'd forgotten we had strictly specified `parse` needs to return the thing.

**Review by akx — APPROVED** [review]
LGTM! Thanks for putting up with the viilaus :)

**Comment by JuhaKiili** [comment]
> LGTM! Thanks for putting up with the viilaus :)

No problem. It was all valid. I was a bit lost with how the serializing worked.
